### PR TITLE
sensors: VehicleIMU improve scheduling

### PR DESCRIPTION
**Pull request by dagar**
- schedule on interval multiple of sensor_gyro publications without increasing the configured integration rate
 - run VehicleIMU WorkItem faster (but still on integer multiples of sensor_gyro) to meet `IMU_INTEG_RATE` rather than raise the integration rate
 - constrain IMU_INTEG_RATE to [100, IMU_GYRO_RATEMAX] (100 Hz is the minimum for ekf2)
 - skip publication interval tracking if gaps in copying sensor_gyro/sensor_accel

In particular this improves the higher rate cases (eg IMU_GYRO_RATEMAX 2-4 kHz). Previously `IMU_INTEG_RATE` would be shifted from 200 Hz -> 250 Hz due to the current sensor_accel/sensor_gyro queue depth (somewhat arbitrary 8). Now in this situation VehicleIMU will still maintain the configured 200 Hz IMU_INTEG_RATE, but instead run twice as fast (400 Hz) to not miss any data.

### CODE REVIEW

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp [hunk 1]
```diff
@@ -364,18 +370,24 @@ void VehicleIMU::UpdateIntegratorConfiguration()
 		_accel_integrator.set_reset_interval(roundf((accel_integral_samples - 0.5f) * _accel_interval.update_interval));
 		_gyro_integrator.set_reset_interval(roundf((gyro_integral_samples - 0.5f) * _gyro_interval.update_interval));
 
-		_sensor_accel_sub.set_required_updates(accel_integral_samples);
-		_sensor_gyro_sub.set_required_updates(gyro_integral_samples);
+		// gyro: find largest integer multiple of gyro_integral_samples
+		for (int n = sensor_gyro_s::ORB_QUEUE_LENGTH; n > 0; n--) {
```
Comment (bkueng) [inline src/modules/sensors/vehicle_imu/VehicleIMU.cpp:374]: I guess the tradeoff is that we use more CPU, in particular if gyro_integral_samples is prime and > Qsize.

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp [hunk 2]
```diff
@@ -129,8 +131,13 @@ bool VehicleIMU::UpdateIntervalAverage(IntervalAverage &intavg, const hrt_abstim
 {
 	bool updated = false;
 
-	if ((intavg.timestamp_sample_last > 0) && (timestamp_sample > intavg.timestamp_sample_last)) {
-		intavg.interval_sum += (timestamp_sample - intavg.timestamp_sample_last);
+	const float interval_us = (timestamp_sample - intavg.timestamp_sample_last);
+	const float imu_integ_interval_us = 1e6f / _param_imu_integ_rate.get();
+
+	if ((intavg.timestamp_sample_last > 0) && (timestamp_sample > intavg.timestamp_sample_last)
+	    && (interval_us <= imu_integ_interval_us)) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp:138]: I was thinking only lockstep SITL would be that low, but I suppose that's not necessarily true. It's mainly to skip and reset the averaging if there's a large gap from some exceptional event like sensor/fifo reset, etc. Let me see what else would be easy.

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp:374]: In the "real" cases that motivated the change it was less CPU overall compared to increasing IMU_INTEG_RATE, which then made ekf2 and the attitude controller(s) run more often.

Example
 - icm20689 at 2000 Hz
 - VehicleIMU runing at 400 Hz to gather all data and publish vehicle_imu at 200 Hz

Now I'm wondering if we should care about the possibility of someone accidentally configuring a pathological case. As a start to limit the scope a bit I'll add typical IMU_INTEG_RATE enum values (likely just 100, 200, 400 Hz) to discourage hand tweaking or at least indicate uncharted territory.

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp [hunk 3]
```diff
@@ -108,10 +108,12 @@ void VehicleIMU::ParametersUpdate(bool force)
...
 		_gyro_calibration.ParametersUpdate();
 
-		// constrain IMU integration time 1-20 milliseconds (50-1000 Hz)
-		int32_t imu_integration_rate_hz = constrain(_param_imu_integ_rate.get(), 50, 1000);
+		// constrain IMU integration time 1-10 milliseconds (100-1000 Hz)
+		int32_t imu_integration_rate_hz = constrain(_param_imu_integ_rate.get(),
+						  100, math::max(_param_imu_gyro_ratemax.get(), 1000));
 
 		if (imu_integration_rate_hz != _param_imu_integ_rate.get()) {
+			PX4_WARN("IMU_INTEG_RATE updated %d -> %d", _param_imu_integ_rate.get(), imu_integration_rate_hz);
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp:116]: Changing the value of a non-volatile parameter silently seemed like bad form.

**Comment by dagar** [comment]
From the test rack.

Holybro Durandal at 4 kHz

Screenshot from 2020-07-08 12-57-01
Screenshot from 2020-07-08 12-57-25

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp (hunk 1 above)
Comment (bkueng) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp:391]: yeah we can leave it at that. I was thinking of (future) IMU's with exotic update rates.

**Pull request merged**